### PR TITLE
FlatUI.Select.Filter adds a form input to select lists for filtering

### PR DESCRIPTION
**Pull request by anutron**
* also re-adds the prod version of the js file to the repo;
  be sure to run `grunt prod` for releases future me!

image

### CODE REVIEW

File: dist/dev/js/behavior-ui-standalone.js [hunk 1]
```diff
@@ -11202,7 +11202,9 @@ window.addEvent('load', function(){
   if (!window.behavior || !behavior.getFilter('FlatUI.Select')) return;
 
   var styleSelect = function(select){
-    if (select.hasBehavior('FlatUI.Select') || select.hasClass('no-flat-select')) return;
```
Comment (anutron) [inline dist/dev/js/behavior-ui-standalone.js]: note: build artifact file

File: dist/dev/js/behavior-ui.js [hunk 2]
```diff
@@ -23656,7 +23656,9 @@ window.addEvent('load', function(){
   if (!window.behavior || !behavior.getFilter('FlatUI.Select')) return;
 
   var styleSelect = function(select){
-    if (select.hasBehavior('FlatUI.Select') || select.hasClass('no-flat-select')) return;
```
Comment (anutron) [inline dist/dev/js/behavior-ui.js]: note: build artifact file

File: js/Source/FlatUI.Select/Behavior.FlatUI.Select.js [hunk 3]
```diff
@@ -14,32 +14,49 @@ provides: [Behavior.FlatUI.Select]
 ...
 */
 
-Behavior.addGlobalFilter('FlatUI.Select', {
+(function(){
```
Comment (anutron) [inline js/Source/FlatUI.Select/Behavior.FlatUI.Select.js]: Only change intended here is the addition of new filter for the filter variant.

File: js/Source/FlatUI.Select/Behavior.FlatUI.Select.js [hunk 4]
```diff
@@ -14,32 +14,49 @@ provides: [Behavior.FlatUI.Select]
...
+
+// the filter for both FlatUI.Select and FlatUI.Select.Filter
+// is identical; the only difference is which class is used
+var init = function(el, api, filterable){
+  var cls = filterable ? FlatUI.Select.Filter : FlatUI.Select;
```
Comment (davywentworth) [inline js/Source/FlatUI.Select/Behavior.FlatUI.Select.js]: i don't know if `cls` is standard in js, but in ruby, we use the word `klass`. it's clearer.

File: js/Source/FlatUI.Select/FlatUI.Select.Filter.js [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+FlatUI.Select.Filter = new Class({
+
+  Extends: FlatUI.Select,
+
+  build: function(){
+    this.parent.apply(this, arguments);
+    // build the DOM elements for our filter, inject them into the list UI
+    this.filterLi = new Element('li.filter').inject(this.list, 'top');
+    this.filterInput = new Element('input.form-control[placeholder=search]');
+    new Element('div.form-group.no-margin.padding-small').adopt(
```
Comment (davywentworth) [inline js/Source/FlatUI.Select/FlatUI.Select.Filter.js]: hmm. not sure about this class inclusion. i think it should either be explicit styling, or a variable that's passed in.

File: js/Source/FlatUI.Select/FlatUI.Select.Filter.js [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+    this.filterInput.set('value', '');
+    this.formFilter.filter();
+  },
+
+  show: function(e){
+    if (this.disabled || this.open) return;
+    this.parent.apply(this, arguments);
+    // on show, select the filter input, so the user can just start typing
+    // delay is required because of the transition
+    this.filterInput.select.delay(300, this.filterInput);
```
Comment (davywentworth) [inline js/Source/FlatUI.Select/FlatUI.Select.Filter.js]: on mobile, we've removed the transition, so this shouldn't be hardcoded.

Comment (davywentworth) [inline js/Source/FlatUI.Select/FlatUI.Select.Filter.js]: also,  the video doesn't show it being pre-selected. sup wit dat?

File: js/Source/FlatUI.Select/FlatUI.Select.Filter.js (hunk 5 above)
Comment (anutron) [inline js/Source/FlatUI.Select/FlatUI.Select.Filter.js]: Good point. I'll make this default but configurable.

File: js/Source/FlatUI.Select/FlatUI.Select.Filter.js (hunk 6 above)
Comment (anutron) [inline js/Source/FlatUI.Select/FlatUI.Select.Filter.js]: I made the video before I added this line I think. Also, hotkeys make no sense on mobile, so I'm going to leave this here. Even if the transition is gone, it still functions well.

**Pull request closed without merging**